src/v/cluster/rm_stm.cc

```diff
@@ -2847,4 +2853,76 @@ ss::future<> rm_stm::clear_old_idempotent_pids() {
     }
 }
 
+std::ostream&
+operator<<(std::ostream& o, const rm_stm::inflight_requests& reqs) {
+    fmt::print(
+      o,
+      "{{ tail: {}, term: {}, request cache size: {} }}",
+      reqs.tail_seq,
+      reqs.term,
+      reqs.cache.size());
+    return o;
+}
+
+std::ostream& operator<<(std::ostream& o, const rm_stm::mem_state& state) {
+    fmt::print(
+      o,
+      "{{ estimated: {}, tx_start: {}, tx_starts: {}, expected: {}, preparing: "
+      "{} }}",
+      state.estimated.size(),
+      state.tx_start.size(),
+      state.tx_starts.size(),
+      state.expected.size(),
+      state.preparing.size());
+    return o;
+}
+
+std::ostream& operator<<(std::ostream& o, const rm_stm::log_state& state) {
+    fmt::print(
+      o,
+      "{{ fence_epochs: {}, ongoing_m: {}, ongoing_set: {}, prepared: {}, "
+      "aborted: {}, abort_indexes: {}, seq_table: {}, tx_seqs: {}, expiration: "
+      "{}}}",
+      state.fence_pid_epoch.size(),
+      state.ongoing_map.size(),
+      state.ongoing_set.size(),
+      state.prepared.size(),
+      state.aborted.size(),
+      state.abort_indexes.size(),
+      state.seq_table.size(),
+      state.tx_seqs.size(),
+      state.expiration.size());
+    return o;
+}
+
+void rm_stm::log_tx_stats() {
+    ssx::background
+      = ssx::spawn_with_gate_then(_gate, [this] {
+            return _state_lock.hold_read_lock()
+              .then([this]([[maybe_unused]] ss::basic_rwlock<>::holder units) {
+                  _ctx_log.info(
+                    "tx root mem_tracker consumption: {}",
+                    human::bytes(
+                      static_cast<double>(_tx_root_tracker.consumption())));
+                  if (unlikely(
+                        _ctx_log.logger().is_enabled(ss::log_level::debug))) {
```

Comment:
perhaps this log level check should be factored out before taking the read lock?